src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Helpers.cs

```diff
@@ -36,37 +35,57 @@ public static partial class JsonSerializer
             return success;
         }
 
-        private static void WriteUsingMetadata<TValue>(Utf8JsonWriter writer, in TValue value, JsonTypeInfo jsonTypeInfo)
+        private static void WriteUsingGeneratedSerializer<TValue>(Utf8JsonWriter writer, in TValue value, JsonTypeInfo jsonTypeInfo)
         {
-            if (jsonTypeInfo is JsonTypeInfo<TValue> typedInfo &&
-                typedInfo.Serialize != null &&
+            if (jsonTypeInfo.HasSerialize &&
+                jsonTypeInfo is JsonTypeInfo<TValue> typedInfo &&
                 typedInfo.Options._context?.CanUseSerializationLogic == true)
             {
+                Debug.Assert(typedInfo.Serialize != null);
                 typedInfo.Serialize(writer, value);
                 writer.Flush();
             }
             else
             {
-                WriteStack state = default;
-                state.Initialize(jsonTypeInfo, supportContinuation: false);
+                WriteUsingSerializer(writer, value, jsonTypeInfo);
+            }
+        }
+
+        private static void WriteUsingSerializer<TValue>(Utf8JsonWriter writer, in TValue value, JsonTypeInfo jsonTypeInfo)
+        {
+            Debug.Assert(jsonTypeInfo is not JsonTypeInfo<TValue>, "Incorrect method called.");
 
-                JsonConverter converter = jsonTypeInfo.PropertyInfoForTypeInfo.ConverterBase;
-                Debug.Assert(converter != null);
+            WriteStack state = default;
+            state.Initialize(jsonTypeInfo, supportContinuation: false);
 
-                Debug.Assert(jsonTypeInfo.Options != null);
+            JsonConverter converter = jsonTypeInfo.PropertyInfoForTypeInfo.ConverterBase;
+            Debug.Assert(converter != null);
+            Debug.Assert(jsonTypeInfo.Options != null);
 
-                WriteCore(converter, writer, value, jsonTypeInfo.Options, ref state);
+            // For performance, the code below is a lifted WriteCore() above.
+            if (converter is JsonConverter<TValue> typedConverter)
```

Comment:
Assuming generic type tests are expensive, is this also an opportunity for us to improve the hot path?

Reply:
@eiriktsarpalis what kind of converter is the most popular? (is it DefaultObjectConverter<T> ?)

Reply:
Possibly, alongside with all the common primitive type converters. Why do you ask?

Reply:
I was just wondering if we could introduce a fast path for the most popular one if it's `sealed`. Cast to sealed classes is almost free.

Reply:
Since this concerns root-level types `ObjectDefaultConverter<T>` would probably be the most common type. If we do try this it out we should make sure it doesn't regress other common root-level converters types like `JsonNodeConverter` or custom converters registered by users.